beacon_node/beacon_chain/src/canonical_head.rs

```diff
@@ -0,0 +1,1307 @@
+        self.cached_head.write()
+    }
+
+    /// Access a read-lock for fork choice.
+    pub fn fork_choice_read_lock(&self) -> RwLockReadGuard<BeaconForkChoice<T>> {
+        self.fork_choice.read()
+    }
+
+    /// Access a write-lock for fork choice.
+    pub fn fork_choice_write_lock(&self) -> RwLockWriteGuard<BeaconForkChoice<T>> {
+        self.fork_choice.write()
+    }
+}
+
+impl<T: BeaconChainTypes> BeaconChain<T> {
+    /// Contains the "best block"; the head of the canonical `BeaconChain`.
+    ///
+    /// It is important to note that the `snapshot.beacon_state` returned may not match the present slot. It
+    /// is the state as it was when the head block was received, which could be some slots prior to
+    /// now.
+    pub fn head(&self) -> CachedHead<T::EthSpec> {
+        self.canonical_head.cached_head()
+    }
+
+    /// Apply a function to an `Arc`-clone of the canonical head snapshot.
+    ///
+    /// This method is a relic from an old implementation where the canonical head was not behind
+    /// an `Arc` and the canonical head lock had to be held whenever it was read. This method is
+    /// fine to be left here, it just seems a bit weird.
+    pub fn with_head<U, E>(
+        &self,
+        f: impl FnOnce(&BeaconSnapshot<T::EthSpec>) -> Result<U, E>,
+    ) -> Result<U, E>
+    where
+        E: From<Error>,
+    {
+        let head_snapshot = self.head_snapshot();
+        f(&head_snapshot)
+    }
+
+    /// Returns the beacon block root at the head of the canonical chain.
+    ///
+    /// See `Self::head` for more information.
+    pub fn head_beacon_block_root(&self) -> Hash256 {
+        self.canonical_head
+            .cached_head_read_lock()
+            .snapshot
+            .beacon_block_root
+    }
+
+    /// Returns the slot of the highest block in the canonical chain.
+    pub fn best_slot(&self) -> Slot {
+        self.canonical_head
+            .cached_head_read_lock()
+            .snapshot
+            .beacon_block
+            .slot()
+    }
+
+    /// Returns a `Arc` of the `BeaconSnapshot` at the head of the canonical chain.
+    ///
+    /// See `Self::head` for more information.
+    pub fn head_snapshot(&self) -> Arc<BeaconSnapshot<T::EthSpec>> {
+        self.canonical_head.cached_head_read_lock().snapshot.clone()
+    }
+
+    /// Returns the beacon block at the head of the canonical chain.
+    ///
+    /// See `Self::head` for more information.
+    pub fn head_beacon_block(&self) -> Arc<SignedBeaconBlock<T::EthSpec>> {
+        self.head_snapshot().beacon_block.clone()
+    }
+
+    /// Returns a clone of the beacon state at the head of the canonical chain.
+    ///
+    /// Cloning the head state is expensive and should generally be avoided outside of tests.
+    ///
+    /// See `Self::head` for more information.
+    pub fn head_beacon_state_cloned(&self) -> BeaconState<T::EthSpec> {
+        // Don't clone whilst holding the read-lock, take an Arc-clone to reduce lock contention.
```

Comment:
this comment could go above in the `head_beacon_block()` function as well - not super necessary but I had spent time thinking through that before getting here :)

Reply:
Oh yeah, good point! I'll just clone the `head_beacon_block` `Arc` directly rather than cloning the `snapshot` `Arc` then cloning the beacon block `Arc`!